src/de/serde.rs

```diff
@@ -659,11 +713,25 @@ impl<'de> de::Deserializer<'de> for Deserializer {
                     iter: doc.into_iter(),
                     value: None,
                     len,
+                    options: self.options,
                 })
             }
         }
     }
 
+    #[inline]
+    fn deserialize_bytes<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        match self.value {
+            Some(Bson::ObjectId(oid)) if !self.is_human_readable() => {
```

Comment:
This special-casing of `ObjectId` seems like it might fit better as a case in `deserialize_any`?

Reply:
The reason this is here is because in the deserialize implementation of `ObjectId`, it checks to see if the format is human readable and, if not, uses `deserialize_bytes` instead of `deserialize_any`. The serializer then uses that as a hint that `ObjectId` is being deserialized. We can't do this in `deserialize_any` because there's no way to hint to the deserializer that an `ObjectId` is being deserialized as opposed to say `Bson`, which needs the `{ "$oid": <hex string> }` format to know its deserializing an ObjectId.